pygfx/renderers/wgpu/wgsl/points.wgsl

```diff
@@ -313,6 +313,75 @@ fn fs_main(varyings: Varyings) -> FragmentOutput {
 }
 
 
+fn get_signed_distance_to_pygfx_outer_triangle(coord_in: vec2<f32>, size: f32) -> vec2<f32> {
+    let m_pi = 3.141592653589793;
+    let m_sqrt_2 = 1.4142135623730951;
+    let half_size = size * 0.5;
+
+    // The coords below assume OpenGL coords, but y is flipped, so we flip the coords.
+    var coord = vec2<f32>(coord_in.x, -coord_in.y);
+
+    // https://math.stackexchange.com/a/4073070
+    // equilateral triangle has length of size
+    //    sqrt(3) - 1
+    let triangle_x = (sqrt(3.) - 1.);
+    let one_minus_triangle_x = 2. - sqrt(3.);
+    let triangle_length = m_sqrt_2 * triangle_x;
+
+    let pygfx_width = 0.10;
+
+    // Offset the coordinates so we can take up the maximum of the quad
+    let coord_1 = coord / size + vec2<f32>(0.5);
+
+    let v1 = normalize(vec2<f32>(one_minus_triangle_x, -1));
+    let r1_out = dot(coord_1, v1);
+    let r1_in  = r1_out + pygfx_width;
+    let r1 = max(r1_out, -r1_in);
+
+    let v2 = normalize(vec2<f32>(-1, one_minus_triangle_x));
+    let r2_out = dot(coord_1, v2);
+    let r2_in  = r2_out + pygfx_width;
+    let r2 = max(r2_out, -r2_in);
+
+    let v3 = normalize(vec2<f32>(triangle_x, triangle_x));
+    let r3_out = dot(coord_1 - vec2(1, one_minus_triangle_x), v3);
+    let r3_in  = r3_out + pygfx_width;
+    let r3 = max(r3_out, -r3_in);
+
+    // return max(max(r1, r2), r3);
+    let outer_edge_triangle = max(
+        max(r1_out, max(r2_out, r3_out)),
+        -max(r1_in, max(r2_in, r3_in))
+    );
```

Comment:
After writing these, I was curious if you were consider ever writting "intersection", "union", "difference" as macros.

It would have made it make more sense to me as writing, then.

Eventually, I got used to it.

Reply:
it depends on how good the compiler is. there's usually a chain of them. in this case the first compilation is handled by naga (WGSL -> shader code), and the second compilation is performed by the GPU and its drivers. In those steps it is possible that the macros simply get compiled away and there will be no performance difference. but it is highly dependent on the versions of naga and drivers, and the specific hardware capabilities. so all in all, using macros may impact performance but it also may not, depending on the environment in which the shader code runs.

Reply:
Interesting.  Thanks!